### PR TITLE
core, p2p: status validation and expiration

### LINKED ISSUE #444: Validate status message and store peer status
## Expected Behavior

The p2p service has a peer manager that stores the validation of peer status messages, clears the validation after a set timeout (unless updated before the set timeout expires), and drops any peers that are sending invalid status message or no status messages at all after a set timeout.

## Current Behavior

The p2p service does not have a peer manager. The validation of compatible status messages is currently being stored on the `host` struct and status messages are not being properly validated.

```
	peerStatus  map[peer.ID]bool // TODO: investigate peer status storage options
```

```
	// TODO: implement status message validation
	case hostStatus.String() == msg.String():
		log.Trace(
			"status match",
			"host", stream.Conn().LocalPeer(),
			"peer", stream.Conn().RemotePeer(),
		)

		// TODO: investigate peer status storage options
		s.host.peerStatus[stream.Conn().RemotePeer()] = true
```

### PR DESCRIPTION
**Pull request by ryanchristo**
## Changes
 
This pull request adds status message validation and status message expiration.

## Tests:
```
go test ./p2p -v
```

### Issues:

closes #444

### CODE REVIEW

File: p2p/discovery.go [hunk 1]
```diff
@@ -42,7 +40,7 @@ type discovery struct {
 	mdns libp2pdiscovery.Service
 }
 
-// newDiscovery creates a new discovery instance from the host
+// newDiscovery creates a new discovery submodule
```
Comment (noot) [inline p2p/discovery.go]: idk if the right word is submodule, I think instance is more correct?

File: p2p/discovery.go [hunk 2]
```diff
@@ -99,17 +97,14 @@ func (d *discovery) startMdns() {
 // HandlePeerFound is event handler called when a peer is found with discovery
 func (n Notifee) HandlePeerFound(p peer.AddrInfo) {
 	log.Trace(
-		"Peer found using mDNS discovery service",
-		"host", n.host.ID(),
+		"Peer found using discovery",
```
Comment (noot) [inline p2p/discovery.go]: being nitpicky but I feel like the "using discovery" is a bit redundant, are there other ways we would find peers?

File: p2p/p2p.go [hunk 3]
```diff
@@ -152,14 +136,27 @@ func (s *Service) receiveCoreMessages() {
...
+			// broadcast message to connected peers
+			s.host.broadcast(msg)
+
+		} else {
+
+			// check if status exchange is enabled
+			if !s.host.noStatus {
+
+				// update host status message
+				s.status.setHostMessage(msg)
```
Comment (noot) [inline p2p/p2p.go]: so just to make sure, this saves the most current status message we get from the network, and then sends that if someone connects?

File: p2p/status.go [hunk 4]
```diff
@@ -19,59 +19,69 @@ package p2p
...
 	"time"
 
-	"github.com/ChainSafe/gossamer/common"
 	log "github.com/ChainSafe/log15"
 	"github.com/libp2p/go-libp2p-core/network"
 	"github.com/libp2p/go-libp2p-core/peer"
 )
 
+// ExpireStatusInterval is the time between checking if status expired
+const ExpireStatusInterval = 20 * time.Minute
```
Comment (noot) [inline p2p/status.go]: is this from somewhere or arbitrary?

File: p2p/discovery.go (hunk 1 above)
Comment (ryanchristo) [reply p2p/discovery.go]: will fix

File: p2p/p2p.go (hunk 3 above)
Comment (ryanchristo) [reply p2p/p2p.go]: correct, this saves the current status message for the host, which will be used when sending status messages to connected peers

File: p2p/discovery.go (hunk 2 above)
Comment (ryanchristo) [reply p2p/discovery.go]: You're right, not sure how else we would "find peers".

File: p2p/status.go (hunk 4 above)
Comment (ryanchristo) [reply p2p/status.go]: arbitrary at the moment, the specification does not specify

File: p2p/discovery.go (hunk 1 above)
Comment (ryanchristo) [reply p2p/discovery.go]: fixed

File: p2p/discovery.go (hunk 2 above)
Comment (ryanchristo) [reply p2p/discovery.go]: fixed

File: p2p/p2p.go [hunk 5]
```diff
@@ -38,27 +37,6 @@ type Service struct {
...
-
-// PeerInfo is network information about peers needed for the rpc server
-type PeerInfo struct {
-	PeerId          string
-	Roles           byte
-	ProtocolVersion uint32
-	BestHash        common.Hash
-	BestNumber      uint64
-}
-
```
Comment (ryanchristo) [inline p2p/p2p.go]: moved to separate file

File: p2p/p2p.go (hunk 3 above)
Comment (ryanchristo) [reply p2p/p2p.go]: the host status message is generated in the core service and sent to the p2p service and then set (or updated) in the status submodule where it is used to validate status messages from peers

File: core/service.go [hunk 6]
```diff
@@ -81,6 +81,9 @@ func NewService(cfg *Config, newBlocks chan types.Block) (*Service, error) {
 // Start starts the core service
 func (s *Service) Start() error {
 
+	// TODO: generate host status message and send to p2p service on startup
+	// msgSend <- hostMessage
```
Comment (ryanchristo) [inline core/service.go]: addressed in #443

File: core/service.go [hunk 7]
```diff
@@ -156,6 +159,12 @@ func (s *Service) handleReceivedBlock(block types.Block) (err error) {
 	// send block announce message to p2p service
 	s.msgSend <- msg
 
+	// TODO: check if host status message needs to be updated based on new block
+	// information, if so, generate host status message and send to p2p service
+
+	// TODO: send updated host status message to p2p service
+	// s.msgSend <- msg
```
Comment (ryanchristo) [inline core/service.go]: addressed in #443

File: p2p/p2p.go (hunk 3 above)
Comment (ryanchristo) [reply p2p/p2p.go]: also peer status messages are never sent to core service, just stored here as well when needing to generate `PeerInfo` for connected peers (`PeerInfo` is now accessible with `p2p.Peers()` which is needed for rpc and runtime)

File: core/service.go (hunk 7 above)
Comment (ryanchristo) [reply core/service.go]: also related to #398

**Comment by ryanchristo** [comment]
Will explore ways to improve expiration following discussion with David.

File: p2p/host.go [hunk 8]
```diff
@@ -253,6 +264,16 @@ func (h *host) peers() []peer.ID {
 	return h.h.Network().Peers()
 }
 
+// peerConnected checks if peer is connected
+func (h *host) peerConnected(peer peer.ID) (connected bool) {
+	for _, p := range h.peers() {
+		if p == peer {
+			connected = true
```
[the remark this answers is not in the document]
Comment (ryanchristo) [reply p2p/host.go]: fixed

File: p2p/status.go [hunk 9]
```diff
@@ -122,10 +119,85 @@ func (status *status) handleMessage(stream network.Stream, msg *StatusMessage) {
...
+		}
+
+		// handle status message expiration
+		go status.expireStatus(ctx, peer)
+
+	} else {
+
+		// update peer status
+		status.peerConfirmed[peer] = time.Time{}
+		status.peerMessage[peer] = nil
```
[the remark this answers is not in the document]
Comment (ryanchristo) [reply p2p/status.go]: fixed

File: p2p/status.go [hunk 10]
```diff
@@ -122,10 +119,85 @@ func (status *status) handleMessage(stream network.Stream, msg *StatusMessage) {
...
+		return false
+	case msg.MinSupportedVersion > status.hostMessage.ProtocolVersion:
+		log.Error("Failed to validate status message", "err", "protocol version")
+		return false
+	}
+	return true
+}
+
+// sendNextMessage sends status messages to the connected peer
+func (status *status) sendNextMessage(ctx context.Context, peer peer.ID) {
```
[the remark this answers is not in the document]
Comment (ryanchristo) [reply p2p/status.go]: fixed

**Comment by ryanchristo** [comment]
Wahoo!

**Pull request merged**